### PR TITLE
PyTorch neural network force fitting

### PR DESCRIPTION
**Pull request by rohskopf**
This PR adds PyTorch NN force fitting using SNAP descriptors. 

We make use of a [recent LAMMPS addition](https://github.com/lammps/lammps/pull/3342) to `compute snap`, which calculates individual bispectrum gradients for each atom. We extract these gradients in FitSNAP just like we do when fitting linear models. The most recent version of LAMMPS is therefore required to run the example, but having older versions of LAMMPS still works if you are not doing nonlinear fitting.

The example in `examples/Ta_PyTorch_NN` has a README explaining how to run the example input script for NN force fitting with PyTorch. There's also a finite difference force test in `examples/Ta_PyTorch_NN/fd_force_check.py`, which uses the same functions in FitSNAP to ensure that our NN calculated forces match those from finite difference. 

**Things to do for later PRs:**

- Add a feature to calculate validation and/or test errors on the fly, like we do with training error.
- Add a function or post-processing script to print the LAMMPS-ready PyTorch models, which was previously done with features in the MLIAP python package. Here we avoid importing MLIAP because of the [current bug with python](https://github.com/lammps/lammps/issues/3204).
- Add NN fitting for multiple element types by creating a separate NN for each atom type.
- Ignore the large descriptor gradient array when fitting only to energies.
- Prune the zero values of the descriptor gradient array for each config, to save space when storing the descriptor gradients for all configs. This is equivalent to pruning with the neighbor list. This should be implemented in `fitsnap3lib/calculators/lammps_snap.py`, following the same procedure in the [new LAMMPS example](https://github.com/lammps/lammps/blob/develop/examples/snap/compute_snap_dgrad.py). 
- Group weights for different configs.

### CODE REVIEW

**Review by mitwood — DISMISSED** [review]
Great new functionality allowing for nonlinear models. The complexities around descriptor derivatives are well handled and extend well to other descriptor sets. Changes are modular and do no disrupt prior capabilities. Users will have to update their LAMMPS library to take advantage of this code.

**Review by mitwood — APPROVED** [review]
